### DISCUSSION TITLE
querying tables in Microsoft .mdb files using python duckdb

### DISCUSSION THREAD

**Discussion by abubelinha** [duckdb/duckdb#13141] [Q&A] [answered]
Hello

Sorry if this is a very off-topic question but in my daily work I need to generate reports from tables stored in a different pretty outdated sources (mainly Microsoft .xls and .mdb files which are edited in old PCs by other people and then shared with me).
I usually connect to .mdb with pyodbc (in the past) or sqlalchemy (now), then query my data with something like this:

```Python
driver = "{Microsoft Access Driver (*.mdb, *.accdb)}"
accessFilePath = "C:\\path\\to\\myAccessDBfile.mdb" 
connstring=(r'DRIVER={{driver}}; DBQ={{accessFilePath}};')
connstring = connstring.replace("{{driver}}",driver)
connstring = connstring.replace("{{accessFilePath}}",accessFilePath)

from sqlalchemy.engine import URL
connection_url = URL.create("access+pyodbc", query={"odbc_connect": connstring})

from sqlalchemy import create_engine
conn = create_engine(connection_url)

import pandas
df1 = pandas.read_csv(...)
df2 = pandas.read_excel(...)

sql = "SELECT * FROM table3;"
df3 = pandas.read_sql(sql,conn)
```
I used to do all sort of stuff merging, filtering and aggregating dataframes with pandas.
Now that I discovered duckdb it is much easier to do all that.

But regarding dataframe 3 ...
1. I don't know how to avoid Pandas for reading my Access .mdb tables.
   Can I somehow reuse `conn` directly from Duckdb?
   Or connect my Access .mdb file in some other way with Duckdb?
2. Also I wonder if I can directly run sql queries joining different tables from .mdb files
3. Or even make updates or inserts in .mdb tables connected from Duckdb

I can't find examples of how to do that.  Do you have any suggestions?
 
Thanks a lot in advance
@abubelinha

**Comment by jgrg** [comment, marked as answer]
DuckDB can connect directly to [SQLite, MySQL and PostgreSQL databases](https://duckdb.org/2024/01/26/multi-database-support-in-duckdb.html), but not ODBC sources.
Presumably you then insert the Pandas dataframe `df3` into DuckDB? Why do you need to eliminate Pandas? I would use your Pandas dataframes to [create or append to tables in DuckDB](https://duckdb.org/docs/guides/python/import_pandas), which you can then run SQL queries to join across.
It can also [load Excel files via the spatial extension](https://duckdb.org/docs/guides/file_formats/excel_import), but I think that only supports `.xlsx` files. So, again, you can use your existing Pandas workflow to populate DuckDB.

**Reply by abubelinha** [reply]
Thanks @jgrg for confirming not-ODBC sources question (I had just asked the same in [this issue](https://github.com/duckdb/duckdb/issues/417#issuecomment-2249895153))

Yes, I would then insert df1, df2, df3 all into duckdb to manipulate my data (instead of using pandas to merge, filter, join, whatever I tried to do with pandas in the past).

After 1 or 2 years using pandas, I just discovered duckdb and liked it a lot because it is quite hard for me to understand pandas syntax (I mean aggregation functions and other selecting/filtering stuff which I was used to do with SQL),
I don't strictly need to eliminate Pandas.  
I was just wondering if it was possible, at least for this kind of duckdb data-loading scripts (just to simplify and speed-up my workflow, as I also read Pandas was quite memory-usage intensive, and I use quite old machines -EDIT: [not only me](https://github.com/duckdb/duckdb/discussions/6645#discussioncomment-10075472)-).

I also was very interested in question 3 since I was googling on how to do that with Pandas (inserting/updating data into Access DB), so I wondered if DuckDB would also make that simpler.

**Comment by Hugoberry** [comment]
I've developed a community extension for ODBC called `nanodbc`. Also, my initial requirement was to connect to MS Access. Give it a try and let me know how it goes.
```sql
install nanodbc from community;
load nanodbc;
from odbc_query(connection='Driver={Microsoft Access Driver (*.mdb, *.accdb)};Dbq=C:\access\files\Northwind 2007.accdb',
  encoding='Window-1252', 
  query='select * from customers');
```
Here is the repo where I could follow up on any feedback https://github.com/Hugoberry/duckdb-nanodbc-extension

**Reply by dmonder** [reply]
How do I get this to work in R? I get the following error in R:

```
Error: HTTP Error: Failed to download extension "nanodbc" at URL "http://community-extensions.duckdb.org/v1.2.2/windows_amd64_mingw/nanodbc.duckdb_extension.gz" (HTTP 403)
```

I can get this to work just fine in DuckDB CLI.

**Reply by abubelinha** [reply]
Thanks @Hugoberry 
I tried it with Python and at a first glance it seems to work well for reading data, but not for writing.
No matter what encoding I use with INSERT queries, the .mdb Access database table always stores garbage characters instead of accents  (see https://github.com/Hugoberry/duckdb-nanodbc-extension/issues/37).
